transport/x/grpc/inbound.go

```diff
@@ -0,0 +1,178 @@
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"errors"
+	"net"
+	"sync"
+
+	"go.uber.org/yarpc/api/transport"
+	internalsync "go.uber.org/yarpc/internal/sync"
+
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
+	"google.golang.org/grpc"
+)
+
+var (
+	errRouterNotSet          = errors.New("router not set")
+	errRouterHasNoProcedures = errors.New("router has no procedures")
+
+	_ transport.Inbound = (*Inbound)(nil)
+)
+
+// Inbound is a grpc transport.Inbound.
+type Inbound struct {
+	once           internalsync.LifecycleOnce
+	lock           sync.Mutex
+	listener       net.Listener
+	inboundOptions *inboundOptions
+	router         transport.Router
+	server         *grpc.Server
+}
+
+// NewInbound returns a new Inbound for the given listener.
+func NewInbound(listener net.Listener, options ...InboundOption) *Inbound {
+	return &Inbound{internalsync.Once(), sync.Mutex{}, listener, newInboundOptions(options), nil, nil}
+}
+
+// Start implements transport.Lifecycle#Start.
+func (i *Inbound) Start() error {
+	return i.once.Start(i.start)
+}
+
+// Stop implements transport.Lifecycle#Stop.
+func (i *Inbound) Stop() error {
+	return i.once.Stop(i.stop)
+}
+
+// IsRunning implements transport.Lifecycle#IsRunning.
+func (i *Inbound) IsRunning() bool {
+	return i.once.IsRunning()
+}
+
+// SetRouter implements transport.Inbound#SetRouter.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.router = router
+}
+
+// Transports implements transport.Inbound#Transports.
+func (i *Inbound) Transports() []transport.Transport {
+	return []transport.Transport{}
+}
+
+func (i *Inbound) start() error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	if i.router == nil {
+		return errRouterNotSet
+	}
+	serviceDescs, err := i.getServiceDescs()
+	if err != nil {
+		return err
+	}
+	server := grpc.NewServer(
+		grpc.CustomCodec(customCodec{}),
+		// TODO: does this actually work for yarpc
+		// this needs a lot of review
+		grpc.UnaryInterceptor(otgrpc.OpenTracingServerInterceptor(i.inboundOptions.getTracer())),
+	)
+	for _, serviceDesc := range serviceDescs {
+		server.RegisterService(serviceDesc, noopGrpcStruct{})
+	}
+	go func() {
+		// TODO there should be some mechanism to block here
+		// there is a race because the listener gets set in the grpc
+		// Server implementation and we should be able to block
+		// until Serve initialization is done
+		//
+		// It would be even better if we could do this outside the
+		// lock in i
+		//
+		// TODO Server always returns a non-nil error but should
+		// we do something with some or all errors?
+		_ = server.Serve(i.listener)
+	}()
+	i.server = server
+	return nil
+}
+
+func (i *Inbound) stop() error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	if i.server != nil {
+		i.server.GracefulStop()
+	}
+	return nil
+}
+
+func (i *Inbound) getServiceDescs() ([]*grpc.ServiceDesc, error) {
+	// TODO: router.Procedures() is not guaranteed to be immutable
```

Comment:
I filed this a while ago on https://github.com/yarpc/yarpc-go/issues/825. I added a comment here. A general note, there is no consistency in yarpc-go about linking to GitHub issues - in our non-generated golang code in yarpc-go, I count 97 TODOs, almost none of which have links to issues. In the spirit of tracking issues, I filed https://github.com/yarpc/yarpc-go/issues/919